Generalised AxesManager iterpath

thomasaarholt opened the pull request:
### Description of the change
I've generalised the `iterpath` property of the AxesManager so that instead of just `"serpentine"` or `"flyback"`, one can now give any iterable or iterator that generates indices of the correct size. This includes _generators_.

This affects the behaviour of `for index in s.axes_manager`, including `m.multifit(iterpath=iterpath)` so that one can specify the order of iteration or which specific positions should be iterated through.

- I've changed from the private property `axes_manager._iterpath` to the `@property` `axes_manager.iterpath`. I'd argue that this doesn't change the API to any significant degree (the multifit argument is unchanged). The new property approach includes checks to ensure that the iterpath is correctly entered.

- I've tidied up the logic around the `__next__` and `__iter__` methods, and removed the `_index` attribute since it isn't used anywhere and can't be used for anything useful.

- Note that since generators and iterators don't have a `len`, `m.multifit` gives a progressbar that is too long. I don't see any good way around that.

- ~~Slightly annoyingly, since I'm relying on np.ndindex and the serpentine generator (which is too complicated for me to reverse engineer), the custom iterpath must follow numpy indexing.~~ ~~This is the reason for the [::-1] in the `AxesManager.__next__` code. Hence, any custom iterpath must also give indices in reverse order (`(1,0,0)` becomes `(0,0,1)`).~~

- I've fixed the above by creating a custom modified `np.ndindex` to reverse its indices.

### Progress of the PR
- [x] Change implemented (can be split into several points),
- [x] update docstring (if appropriate),
- [x] update user guide pending merge of #2549
- [x] add tests,
- [ ] ready for review.

### Minimal example of the bug fix or the new feature
```python
import hyperspy.api as hs
import numpy as np

s = hs.signals.Signal1D(np.ones((3, 3, 10)))
ax = s.axes_manager
m = s.create_model()
G = hs.model.components1D.Gaussian()
m.append(G)

# Example 1
indices = np.array([(0,0), (0,1), (2,2)])
s.axes_manager.iterpath = indices
for i in s.axes_manager:
    print(i)

# Example 2
m.multifit(iterpath=indices)

# Example 3 (raises error)
s.axes_manager.iterpath = "test"
```

thomasaarholt committed:
generalised the iterpath property

thomasaarholt committed:
fix test

thomasaarholt committed:
fix failing test

thomasaarholt committed:
revert a change, custom iterpath must follow numpy indexing, not hyperspy

thomasaarholt committed:
fix reversed custom iterpath issue

thomasaarholt committed:
much simpler ndindex implementation

thomasaarholt committed:
now takes shape in hyperspy order

thomasaarholt committed:
full coverage on tests now

thomasaarholt committed:
one more coverage bit

thomasaarholt commented:
This PR is ready, but the final commits (docs in the new `axes.rst`) lie in my branch based off `non-uniform axes` [here](https://github.com/thomasaarholt/hyperspy/tree/general_iterpath_non_uniform). I'm keeping these two separate until #2549 is merged in case there are significant changes, so that it's easy to see (if anything) has changed.

ericpre commented on hyperspy/axes.py:
```diff
@@ -649,6 +654,22 @@ def serpentine_iter(shape):
+
+    return ndindex_reversed(shape)
+
+iterpath_error = '''The iterpath scan pattern is set to "{}". \
```
Move this to where it is used in the namespace tidy?

ericpre commented on hyperspy/_signals/signal1d.py:
```diff
@@ -275,6 +275,7 @@ def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
         if self.axes_manager.signal_dimension != 1:
             self.axes_manager.set_signal_dimension(1)
+        self.axes_manager.iterpath = self.axes_manager._iterpath
```
Why is it necessary?

thomasaarholt replied:
Fixed.
The [`test_signal_iterator`](https://github.com/hyperspy/hyperspy/blob/e4c6994b61837a2c54c963d2b1afe3b86c1bf7c4/hyperspy/tests/signal/test_tools.py#L46) test was originally failing without those lines, but at a closer look I believe I can safely modify the test so it is no longer failing. It was calling `next(s).data[0]`, without calling the `__iter__()` method first, which wasn't initialising the `iterpath_generator`.

ericpre reviewed (changes requested):
The docstring of the `AxesManager` needs to be updated.

One suggestion I am not sure: now that `iterpath` is public, maybe the attribute name should be changed to avoid using an abbreviation and be more explicit. If so, may be worth considering changing the corresponding argument of `multifit`.

thomasaarholt committed:
@ericpre comments

thomasaarholt commented:
Thanks for the comments @ericpre, the lines you commented about were annoying me, and I learnt a bit more about `__iter__` and `__next__` at the same time as fixing them.

I'd be happy to change the name - I agree that being clear is good here - although I think I can only suggest `iteration_path`. How does that sound?

Edit: Forgot to change AxesManager docstring. Doing so now.

thomasaarholt committed:
add docstring

ericpre commented on hyperspy/axes.py:
```diff
@@ -961,57 +977,56 @@ def _update_max_index(self):
+            except TypeError:
+                raise TypeError(
+                f"The iterpath '{self._iterpath}' is not a correct iterpath. "
+                "Ensure it is an iterable delivering incides with length equal "
```
The error message doesn't match what it is checked and it would be good to check it.

thomasaarholt replied:
Fixed.

thomasaarholt replied:
In the case of the `self._iterpath` being a generator, I don't think we can check if the length of the indices yielded are the correct size without using up the first generated value. I've added checks for the case where it is not a generator (but an array, list of tuples, etc).

ericpre commented on hyperspy/axes.py:
```diff
@@ -649,6 +654,16 @@ def serpentine_iter(shape):
         else:  # pragma: no cover
             break
 
+def flyback_iter(shape):
```
Make this function private?

ericpre commented on hyperspy/axes.py:
```diff
@@ -773,12 +792,9 @@ def __init__(self, axes_list):
             # Default to Signal1D view if the view is not fully defined
             self.set_signal_dimension(len(axes_list))
 
+        self._iterpath = 'flyback'
```
Should it be `self.iterpath = 'flyback'`?

thomasaarholt replied:
If I move it to the end of the AxesManager `__init__`, then yes :) Done. (Otherwise the `navigation_shape` which it uses, hasn't been created yet.)

ericpre commented on hyperspy/axes.py:
```diff
@@ -961,57 +977,56 @@ def _update_max_index(self):
+                raise ValueError(iterpath_error)
+        else:
+            # Pass a custom indices iterator
+            self._iterpath = path
```
`self._iterpath` should not be set at this stage in case there is an exception below.

thomasaarholt replied:
Thanks, fixed.

ericpre commented on hyperspy/axes.py:
```diff
@@ -679,6 +694,10 @@ class AxesManager(t.HasTraits):
+    iterpath : string or iterable
+        Sets the order of iterating through the indices in the navigation dimension.
+        indices. Can be either "flyback" or "serpentine", or an iterable
+        of hyperspy indices.
```
`hyperspy indices` -> `navigation indices`?

ericpre reviewed (changes requested):
> - Note that since generators and iterators don't have a `len`, `m.multifit` gives a progressbar that is too long. I don't see any good way around that.

When using an custom iterpath, such as an iterable, the length can be known from the length of `_iterpath`.

thomasaarholt committed:
fix ericpre comments and add and fix tests

thomasaarholt committed:
Move and change _iterpath -> iterpath

thomasaarholt commented:
I've fixed your comments. On the topic of changing the name of `iterpath`: @francisco-dlp you requested this feature in the first place 😄 any thoughts on whether we should use a more friendly name? Suggestions? (others welcome too)

thomasaarholt committed:
fix typo

thomasaarholt commented:
> 
> 
> > * Note that since generators and iterators don't have a `len`, `m.multifit` gives a progressbar that is too long. I don't see any good way around that.
> 
> When using an custom iterpath, such as an iterable, the length can be known from the length of `_iterpath`.

In the case of the `self._iterpath` being a generator, I don't think we can check this. We could offer the user the option to add their own length? I just realised, I have to check if the too-long progressbar (this isn't really a problem) can also be too-short (this is a problem), and possibly stop fitting if the iterpath is longer than the number of navigation size.

ericpre commented:
> > > * Note that since generators and iterators don't have a `len`, `m.multifit` gives a progressbar that is too long. I don't see any good way around that.
> > 
> > 
> > When using an custom iterpath, such as an iterable, the length can be known from the length of `_iterpath`.
> 
> In the case of the `self._iterpath` being a generator, I don't think we can check this. We could offer the user the option to add their own length? I just realised, I have to check if the too-long progressbar (this isn't really a problem) can also be too-short (this is a problem), and possibly stop fitting if the iterpath is longer than the number of navigation size.

Since not all iterables are generators and some of them have a length, it would be good using it when possible, for example in the case of:
https://github.com/thomasaarholt/hyperspy/blob/6d496e5bca2126adae57088f96c3723d78c01fa7/hyperspy/tests/axes/test_axes_manager.py#L405-L408

thomasaarholt committed:
check if progressbar length is known

jlaehne commented:
> This PR is ready, but the final commits (docs in the new `axes.rst`) lie in my branch based off `non-uniform axes` [here](https://github.com/thomasaarholt/hyperspy/tree/general_iterpath_non_uniform). I'm keeping these two separate until #2549 is merged in case there are significant changes, so that it's easy to see (if anything) has changed.

The relevant changes to `axes.rst` should now be merged into `non_uniform_axes` through #2304 (#2549 was just based off this PR to include some changes made there -- I now rebased the branch to make that clear)

thomasaarholt committed:
Add GeneratorLen, more improvements

thomasaarholt committed:
add documentation

thomasaarholt commented:
I've rebased onto non_uniform_axes!

Once tests have finished running, I'd appreciate in particular if someone could take a look through the documentation I've added and see if it makes sense. The failing TestSpikesRemovalToolZLP::test_spikes_diagnosis is failing upstream too.

Reminder to self: There is a bit of text in the non_uniform_axes docs that exists both in the new axes.rst and getting_started.rst. I'll see if I can whip up a PR tomorrow evening.

ericpre commented on hyperspy/axes.py:
```diff
@@ -1427,57 +1445,103 @@ def _update_max_index(self):
+            try:
+                maxval = len(self.iterpath)
+                if masked_elements:
+                    # Checking if mask indices exist in the iterpath could take a long time,
+                    # or may not be possible in the case of a generator.
+                    warnings.warn(
+                    ("The progressbar length cannot be estimated when using both custom iterpath and a mask."
+                    "The progressbar may terminate before it appears complete. This can safely be ignored."),
+                    UserWarning
+                    )
+            except TypeError:
+                # progressbar is shown, so user can monitor "iterations per second"
+                # but the length of the bar is unknown
+                maxval = None
+                warnings.warn(
+                    ("The AxesManager `iterpath` is missing the `__len__` method, so does not have a known length. "
+                    "The progressbar will only show run time and iterations per second, but no actual progress indicator."),
+                    UserWarning
+                    )
```
If this is intended behaviour, this shouldn't a warning and adding it as info in the logger should be enough?

thomasaarholt replied:
Yes, I agree with that.

ericpre reviewed (commented):
I have one last comment, other than that, I think it should be good?

thomasaarholt committed:
warning -> logging

thomasaarholt commented:
Thanks @ericpre. Pushed that change now.

ericpre commented:
@thomasaarholt, is it finished from your side?

thomasaarholt commented:
Yes, this can be merged.
